src/e2e/sdk_actions.ts

```diff
@@ -0,0 +1,397 @@
+		// assets.Tokens => non fungible tokens (NFT) of the ERC721 type.
+		//
+		// Here we are creating `Assets` containing some ETH and PRN which will be
+		// used for depositing in the next step.
+		const depositBal = new Assets(
+			{
+				token: assets.ETHZERO,
+				asset: new assets.Amount(ethAmount),
+			},
+			{
+				token: PERUN_ADDR,
+				asset: new assets.Amount(prnAmount),
+			},
+		);
+
+		const { stages, numStages: _numStages } = await session.deposit(
+			depositBal,
+		);
+		// Depositing is a multi-stage process. ERC20 tokens like PRN have to be
+		// approved first, before being transferred. Here this results in the
+		// following stages:
+		//
+		// ETH => One transfer stage.
+		// PRN => One approve stage + One transfer stage.
+		//
+		// console.info("Number of stages: ", _numStages); // > Number of stages: 3
+
+		// One can use this information to update users on the current progress of
+		// onchain transactions since they tend to take up quite some time in the
+		// real world.
+		for await (const [_name, stage] of stages) {
+			// Each stage of a transaction comes with a name which also can be used
+			// for improved UX or logging.
+			const ctx = await stage.wait();
+			// Also it is always advisable to assert the contract transactions status
+			// to be a success.
+			if (ctx.status !== 0x1) {
+				// handle contract error.
+			}
+		}
+	},
+
+	offchainTransfer: async (
+		alice: Session,
+		bob: Session,
+		prnAmount: bigint,
+	): Promise<void> => {
+		const amount = new Assets({
+			token: PERUN_ADDR,
+			asset: new assets.Amount(prnAmount),
+		});
+		// All off-chain transactions return a `TxReceipt`. When a user is the
+		// source for a transaction, e.g. calling `session.transferTo(...)` then
+		// the return value for that call can be awaited, resulting in the
+		// TxReceipt for this transaction.
+		//
+		// When a user is the target for a transaction he will also receive a
+		// txreceipt. These receipts can be caught using the `"receipt"`
+		// eventhandler as shown below.
+		const aliceTxReceipt = await alice.transferTo(amount, bob.address);
+
+		aliceTxReceipt.tx.sender.equals(alice.address); // true
+
+		// We expect BOB to receive a receipt, so we resolve and clear the
+		// timeout as soon as that receipt is received.
+		//
+		// Note how we use a `once` call. If a persistent eventhandler shall be
+		// registered one can just use the equivalent `on` method on the
+		// `ErdstallClient`/`ErdstallSession`.
+		//
+		// Depending on the `subscribe` call issued beforehand bob could also
+		// receive tx receipts for other users. This is the default for an
+		// `ErdstallClient` because the `ErdstallClient` only contains a generic
+		// `subscribe()` call and has no associated address!
+		//
+		// Remember that in the beginning we did a `client.subscribeSelf()`, so it
+		// is assured that the eventhandlers only trigger if we (here it is bob)
+		// are the target.
+		bob.once("receipt", (bobTxReceipt) => {
```

Comment:
Normally yes but since this handler does not really do anything it works. I did it this way because it felt _more natural_ to write it this way. I will add a remark that Bob obviously would have had to set his EventHandler **before** Alice sent her TX.